File: decode/js/rosetta-image-decode/src/decode.ts
import { decodeBmp } from "./internal/bmp.js";
import { decodeGif } from "./internal/gif.js";
import { decodeHeic } from "./internal/heic.js";
import { decodeJpeg } from "./internal/jpeg.js";
import { decodePng } from "./internal/png.js";
import { decodeTiff } from "./internal/tiff.js";
import { decodeWebp } from "./internal/webp.js";
import { DecodeError } from "./errors.js";
import type { DecodedImage, Format } from "./types.js";

export async function decode(bytes: Uint8Array): Promise<DecodedImage> {
  const fmt = detectFormat(bytes);
  if (fmt == null) {
    throw new DecodeError("unsupportedFormat", null, "");
  }
  switch (fmt) {
    case "bmp":
      return decodeBmp(bytes);
    case "png":
      return decodePng(bytes);
    case "gif":
      return decodeGif(bytes);
    case "jpeg":
      return await decodeJpeg(bytes);
    case "webp":
      return await decodeWebp(bytes);
    case "tiff":
      return await decodeTiff(bytes);
    case "heic":
      return await decodeHeic(bytes);
    default:
      throw new DecodeError("unsupportedFormat", fmt, "");
  }
}

export function detectFormat(bytes: Uint8Array): Format | null {
  if (bytes.length < 2) return null;
  if (bytes[0] === 0x42 && bytes[1] === 0x4d) return "bmp";
  if (
    bytes.length >= 8 &&
    bytes[0] === 0x89 && bytes[1] === 0x50 && bytes[2] === 0x4e && bytes[3] === 0x47 &&
    bytes[4] === 0x0d && bytes[5] === 0x0a && bytes[6] === 0x1a && bytes[7] === 0x0a
  ) {
    return "png";
  }
  if (
    bytes.length >= 6 &&
    bytes[0] === 0x47 && bytes[1] === 0x49 && bytes[2] === 0x46 && bytes[3] === 0x38 &&
    (bytes[4] === 0x37 || bytes[4] === 0x39) && bytes[5] === 0x61
  ) {
    return "gif";
  }
  if (bytes.length >= 2 && bytes[0] === 0xff && bytes[1] === 0xd8) {
    return "jpeg";
  }
  if (
    bytes.length >= 12 &&
    bytes[0] === 0x52 && bytes[1] === 0x49 && bytes[2] === 0x46 && bytes[3] === 0x46 &&
    bytes[8] === 0x57 && bytes[9] === 0x45 && bytes[10] === 0x42 && bytes[11] === 0x50
  ) {
    return "webp";
  }
  if (
    bytes.length >= 4 && (
      // Little-endian TIFF: II + 42
      (bytes[0] === 0x49 && bytes[1] === 0x49 && bytes[2] === 0x2a && bytes[3] === 0x00) ||
      // Big-endian TIFF: MM + 42
      (bytes[0] === 0x4d && bytes[1] === 0x4d && bytes[2] === 0x00 && bytes[3] === 0x2a)
    )
  ) {
    return "tiff";
  }
  // HEIC/HEIF: ISO Base Media File Format with ftyp box at bytes 4..7
  // Bytes 8..11 = major brand; only accept HEVC-based brands, reject avif.
  if (
    bytes.length >= 12 &&
    bytes[4] === 0x66 && bytes[5] === 0x74 && bytes[6] === 0x79 && bytes[7] === 0x70
  ) {
    const brand = String.fromCharCode(bytes[8]!, bytes[9]!, bytes[10]!, bytes[11]!);
    if (["heic", "heix", "mif1", "msf1", "hevc", "hevx"].includes(brand)) {
      return "heic";
    }
    // avif and other unrecognized ftyp brands → unsupportedFormat
    return null;
  }
  return null;
}

export function supportedFormats(): Format[] {
  return ["bmp", "png", "gif", "jpeg", "webp", "tiff", "heic"];
}

File: decode/rust/rosetta-image-decode/src/decode.rs
use crate::bmp::decode_bmp;
use crate::gif::decode_gif;
use crate::png::decode_png;
use crate::error::{DecodeError, DecodeErrorKind};
use crate::types::{DecodedImage, Format};
use crate::webp::decode_webp;

/// Decode auto-detects the format from magic bytes and decodes.
pub fn decode(bytes: &[u8]) -> Result<DecodedImage, DecodeError> {
    let fmt = detect_format(bytes).ok_or_else(|| {
        DecodeError::new(DecodeErrorKind::UnsupportedFormat, None, "")
    })?;
    match fmt {
        Format::Bmp => decode_bmp(bytes),
        Format::Gif => decode_gif(bytes),
        Format::Jpeg => crate::jpeg::decode_jpeg(bytes),
        Format::Png => decode_png(bytes),
        Format::Webp => decode_webp(bytes),
        other => Err(DecodeError::new(DecodeErrorKind::UnsupportedFormat, Some(other), "")),
    }
}

/// Returns the Format if the magic bytes match, else None.
pub fn detect_format(bytes: &[u8]) -> Option<Format> {
    if bytes.len() < 2 {
        return None;
    }
    if bytes[0] == 0x42 && bytes[1] == 0x4D {
        return Some(Format::Bmp);
    }
    if bytes.len() >= 8
        && bytes[0] == 0x89
        && bytes[1] == 0x50
        && bytes[2] == 0x4E
        && bytes[3] == 0x47
        && bytes[4] == 0x0D
        && bytes[5] == 0x0A
        && bytes[6] == 0x1A
        && bytes[7] == 0x0A
    {
        return Some(Format::Png);
    }
    // GIF87a or GIF89a
    if bytes.len() >= 6
        && bytes[0] == b'G'
        && bytes[1] == b'I'
        && bytes[2] == b'F'
        && bytes[3] == b'8'
        && (bytes[4] == b'7' || bytes[4] == b'9')
        && bytes[5] == b'a'
    {
        return Some(Format::Gif);
    }
    // JPEG: starts with FF D8 (SOI marker); third byte FF is typical but not required for detection
    if bytes.len() >= 2
        && bytes[0] == 0xFF
        && bytes[1] == 0xD8
    {
        return Some(Format::Jpeg);
    }
    // WebP: RIFF????WEBP (bytes 0-3 = "RIFF", bytes 8-11 = "WEBP")
    if bytes.len() >= 12
        && bytes[0..4] == [b'R', b'I', b'F', b'F']
        && bytes[8..12] == [b'W', b'E', b'B', b'P']
    {
        return Some(Format::Webp);
    }
    None
}

/// Returns the list of formats this port can decode.
pub fn supported_formats() -> Vec<Format> {
    vec![Format::Bmp, Format::Png, Format::Gif, Format::Jpeg, Format::Webp]
}
